Sources/MongoSwift/BSON/BSONError.swift

```diff
@@ -0,0 +1,114 @@
+import Foundation
+
+/// An empty protocol for encapsulating all errors that BSON package can throw.
+public protocol BSONErrorProtocol: LocalizedError {}
+
+/// A protocol describing errors caused by improper usage of the BSON library by the user.
+public protocol BSONUserError: BSONErrorProtocol {}
+
+/// The possible errors that can occur unexpectedly BSON library-side.
+public protocol BSONRuntimeError: BSONErrorProtocol {}
+
+/// Namespace containing all the error types introduced by this BSON library and their dependent types.
+public enum BSONError {
+    /// An error thrown when the user passes in invalid arguments to a BSON method.
+    public struct InvalidArgumentError: BSONUserError {
+        internal let message: String
+
+        public var errorDescription: String? { self.message }
+    }
+
+    /// An error thrown when the BSON library encounters a internal error not caused by the user.
+    /// This is usually indicative of a bug in the BSON library or system related failure.
+    public struct InternalError: BSONRuntimeError {
+        internal let message: String
+
+        public var errorDescription: String? { self.message }
+    }
+
+    /// An error thrown when the BSON library is incorrectly used.
+    public struct LogicError: BSONUserError {
+        internal let message: String
+
+        public var errorDescription: String? { self.message }
+    }
+}
+
+internal func bsonTooLargeError(value: BSONValue, forKey: String) -> BSONErrorProtocol {
```

Comment:
Since this error could reasonably be caused by the user, maybe it makes sense to define a specific error for it (e.g. `BSONDocumentTooLarge` or `BSONDocumentExceededSizeLimit` or something. `InternalError` should generally be used for system issues (e.g. out of memory causes allocation to fail) or invariant failures (i.e. bugs in the BSON library).

Reply:
Sure I agree, I think I'll still wanna add a catch and translate for this new error what do you think is the best MongoError to translate to?

Reply:
I think `MongoError.InternalError` still works fine probably, since if we're creating the document in the driver and make it too big, it's probably our fault.

Reply:
Indeed, I also added a generic BSONErrorProtocol catcher for future potential bson errors:

https://github.com/mongodb/mongo-swift-driver/blob/30fe38686000a37f7c18eb13dd1b711162564578/Sources/MongoSwift/BSON/BSONUtil.swift#L16-L18

Reply:
also fine with me